sklearn/mixture/gaussian_mixture.py

```diff
@@ -0,0 +1,651 @@
+
+    reg_covar : float
+
+    Returns
+    -------
+    covariances : array, shape (n_features, n_features)
+    """
+    avg_X2 = np.dot(X.T, X)
+    avg_means2 = np.dot(nk * means.T, means)
+    covariances = avg_X2 - avg_means2
+    covariances /= X.shape[0]
+    covariances.flat[::len(covariances) + 1] += reg_covar
+    return covariances
+
+
+def _estimate_gaussian_covariance_diag(resp, X, nk, means, reg_covar):
+    """Estimate the diagonal covariance matrices.
+
+    Parameters
+    ----------
+    responsibilities : array-like, shape (n_samples, n_components)
+
+    X : array-like, shape (n_samples, n_features)
+
+    nk : array-like, shape (n_components,)
+
+    means : array-like, shape (n_components, n_features)
+
+    reg_covar : float
+
+    Returns
+    -------
+    covariances : array, shape (n_components, n_features)
+    """
+    avg_X2 = np.dot(resp.T, X * X) / nk[:, np.newaxis]
+    avg_means2 = means ** 2
+    avg_X_means = means * np.dot(resp.T, X) / nk[:, np.newaxis]
+    return avg_X2 - 2 * avg_X_means + avg_means2 + reg_covar
+
+
+def _estimate_gaussian_covariance_spherical(resp, X, nk, means, reg_covar):
+    """Estimate the spherical covariance matrices.
+
+    Parameters
+    ----------
+    responsibilities : array-like, shape (n_samples, n_components)
+
+    X : array-like, shape (n_samples, n_features)
+
+    nk : array-like, shape (n_components,)
+
+    means : array-like, shape (n_components, n_features)
+
+    reg_covar : float
+
+    Returns
+    -------
+    covariances : array, shape (n_components,)
+    """
+    covariances = _estimate_gaussian_covariance_diag(resp, X, nk, means,
+                                                     reg_covar)
+    return covariances.mean(axis=1)
+
+
+def _estimate_gaussian_parameters(X, resp, reg_covar, covariance_type):
+    """Estimate the Gaussian distribution parameters.
+
+    Parameters
+    ----------
+    X : array-like, shape (n_samples, n_features)
+        The input data array.
+
+    resp : array-like, shape (n_samples, n_features)
+        The responsibilities for each data sample in X.
+
+    reg_covar : float
+        The regularization added to each covariance matrices.
+
+    covariance_type : {'full', 'tied', 'diag', 'spherical'}
+        The type of covariance matrices.
+
+    Returns
+    -------
+    nk : array, shape (n_components,)
+        The numbers of data samples in the current components.
+
+    means : array, shape (n_components, n_features)
+        The centers of the current components.
+
+    covariances : array
+        The sample covariances of the current components.
+        The shape depends of the covariance_type.
+    """
+    compute_covariance = {
+        "full": _estimate_gaussian_covariance_full,
+        "tied": _estimate_gaussian_covariance_tied,
+        "diag": _estimate_gaussian_covariance_diag,
+        "spherical": _estimate_gaussian_covariance_spherical}
+
+    nk = resp.sum(axis=0) + 10 * np.finfo(float).eps
```

Comment:
It's better to use `np.finfo(resp.dtype).eps` to make the code adaptive to a change of precision in `dtype`.